<launch>
  <arg name="uav_name" default="$(optenv UAV_NAME uav)"/>

  <arg name="standalone" default="false"/>

  <arg unless="$(arg standalone)" name="nodelet" value="load"/>
  <arg     if="$(arg standalone)" name="nodelet" value="standalone"/>
  <arg unless="$(arg standalone)" name="nodelet_manager" value="$(arg uav_name)_uvdar_nodelet_manager"/>
  <arg     if="$(arg standalone)" name="nodelet_manager" value=""/>

  <arg name="threshold" default="100"/>

  <arg name="left" default="$(optenv BLUEFOX_UV_LEFT)"/>
  <arg name="right" default="$(optenv BLUEFOX_UV_RIGHT)"/>
  <arg name="left_camera_name" default="mv_$(arg left)"/>
  <arg name="right_camera_name" default="mv_$(arg right)"/>
  <arg name="expose_us_left" default="$(optenv BLUEFOX_UV_LEFT_EXPOSE_US)"/>
  <arg name="expose_us_right" default="$(optenv BLUEFOX_UV_RIGHT_EXPOSE_US)"/>

  <arg name="calibrations_folder" default="$(find mrs_uav_general)/config/uvdar_calibrations"/>

  <arg name="blink_process_rate" default="10"/>

  <arg name="accumulator_length"         default="23"/>
  <arg name="pitch_steps"         default="16"/>
  <arg name="yaw_steps"         default="16"/>
  <arg name="max_pixel_shift"         default="3"/>

  <arg name="filterDistLength"         default="10"/>
  <arg name="filterOrientationLength"         default="10"/>

  <!-- Node Settings -->
  <arg name="output" default="screen"/>
  <arg name="proc" default="false"/>
  <arg name="view" default="false"/>
  <arg name="calib" default="false"/>

  <arg name="debug" default="false"/>
  <arg name="visual_debug" default="false"/>
  <arg name="gui" default="false"/>
  <!-- <arg name="gui" default="false"/> -->

  <arg name="publish" default="true"/>

  <arg name="frameNatoName" default="$(optenv NATO_NAME)"/>

  <arg name="beacon" default="false"/>

  <arg name="frequency1" value="6"/>
  <arg name="frequency2" value="10"/>
  <arg name="frequency3" value="15"/>
  <arg name="frequency4" value="30"/>

  <!-- ideal angle: 1.221730476 - pi/2 -->

      <!-- <origin xyz="0.108 0 0.085" rpy="1.57079632679 4.71238898038 1.57079632679" /> -->
  <!--Usage: static_transform_publisher x y z yaw pitch roll frame_id child_frame_id  period(milliseconds)-->
<!-- OR -->
  <group ns="$(arg uav_name)">
    <include file="$(arg calibrations_folder)/extrinsics/$(arg frameNatoName)_left.xml" pass_all_args="true" />
    <include file="$(arg calibrations_folder)/extrinsics/$(arg frameNatoName)_right.xml" pass_all_args="true" />

    <node pkg="nodelet" type="nodelet" name="$(arg uav_name)_uvdar_nodelet_manager" args="manager" output="screen" >
    <!-- <node pkg="nodelet" type="nodelet" name="$(arg uav_name)_uvdar_nodelet_manager" args="manager" output="screen" launch-prefix="debug_roslaunch"> -->
      <param name="num_worker_threads" value="8" />
    </node>

    <node pkg="nodelet" type="nodelet" name="uvdar_bluefox_left" args="$(arg nodelet) bluefox2/SingleNodelet $(arg nodelet_manager)" respawn="true" output="screen" launch-prefix="bash -c 'sleep 5; $0 $@'"> <param name="camera_name" type="string" value="$(arg left_camera_name)"/>
      <param name="identifier" type="string" value="$(arg left)"/>

      <param name="frame_id" type="string" value="$(arg uav_name)/uvcam_left"/>
      <param name="fps" type="double" value="100"/>
      <param name="idpf" type="int" value="0"/>
      <param name="aec" type="bool" value="false"/>
      <param name="expose_us" type="int" value="$(arg expose_us_left)"/>
      <param name="agc" type="bool" value="false"/>
      <param name="gain_db" type="double" value="0.0"/>
      <param name="cbm" type="int" value="0"/>
      <param name="ctm" type="int" value="1"/>
      <param name="dcfm" type="int" value="0"/>
      <param name="hdr" type="bool" value="false"/>
      <param name="wbp" type="int" value="-1"/>
      <param name="request" type="int" value="0"/>
      <param name="mm" type="int" value="0"/>

      <param name="image_raw/compressed/jpeg_quality" type="int" value="90"/>
      <param name="image_raw/theora/keyframe_frequency" value="60" />
      <param name="image_raw/theora/target_bitrate" value="50000" />
      <param name="image_raw/theora/quality" value="8" />
      <param name="image_raw/theora/optimize_for" value="0" />
    </node>

    <node pkg="nodelet" type="nodelet" name="uvdar_bluefox_right" args="$(arg nodelet) bluefox2/SingleNodelet $(arg nodelet_manager)" respawn="true" output="screen" launch-prefix="bash -c 'sleep 7; $0 $@'">
      <param name="camera_name" type="string" value="$(arg right_camera_name)"/>
      <param name="identifier" type="string" value="$(arg right)"/>

      <param name="frame_id" type="string" value="$(arg uav_name)/uvcam_right"/>
      <param name="fps" type="double" value="100"/>
      <param name="idpf" type="int" value="0"/>
      <param name="aec" type="bool" value="false"/>
      <param name="expose_us" type="int" value="$(arg expose_us_right)"/>
      <param name="agc" type="bool" value="false"/>
      <param name="gain_db" type="double" value="0.0"/>
      <param name="cbm" type="int" value="0"/>
      <param name="ctm" type="int" value="1"/>
      <param name="dcfm" type="int" value="0"/>
      <param name="hdr" type="bool" value="false"/>
      <param name="wbp" type="int" value="-1"/>
      <param name="request" type="int" value="0"/>
      <param name="mm" type="int" value="0"/>

      <param name="image_raw/compressed/jpeg_quality" type="int" value="90"/>
      <param name="image_raw/theora/keyframe_frequency" value="60" />
      <param name="image_raw/theora/target_bitrate" value="50000" />
      <param name="image_raw/theora/quality" value="8" />
      <param name="image_raw/theora/optimize_for" value="0" />
    </node>



    <node name="uv_detect" pkg="nodelet" type="nodelet" args="$(arg nodelet) uvdar/UVDARDetector $(arg nodelet_manager)" output="screen" respawn="false"> 
    <!-- <node name="uv_detect" pkg="nodelet" type="nodelet" args="$(arg nodelet) uvdar/UVDARDetector $(arg nodelet_manager)" output="screen" respawn="true" launch-prefix="debug_roslaunch"> --> 
      <!-- <node name="uv_detect" pkg="uvdar" type="uv_detector_node" output="screen" launch-prefix="urxvt -e gdb -q -x /home/viktor/gdb.cmds -/-args"> -->
        <!-- <node name="uv_detect" pkg="nodelet" type="nodelet" args="$(arg nodelet) uvdar/UVDARDetector $(arg nodelet_manager)" output="screen" respawn="true" launch-prefix="valgrind -/-tool=callgrind -/-callgrind-out-file=/home/viktor/callgrind.out -/-instr-atstart=no -/-collect-atstart=yes"> --> 

      <param name="uav_name"		type="string" value="$(arg uav_name)"/> 
      <param name="debug"		type="bool" value="$(arg debug)"/> 
      <!-- <param name="gui"             type="bool" value="true"/> -->
      <param name="gui"             type="bool" value="false"/>
      <!-- <param name="publish_visualization"             type="bool" value="true"/> -->
      <param name="publish_visualization"             type="bool" value="true"/>
      <param name="publish"         type="bool" value="$(arg publish)"/>
      <param name="justReport" type="bool" value="true"/>
      <param name="threshold" type="int" value="$(arg threshold)"/>

      <rosparam param="camera_topics"> ["camera_left", "camera_right"] </rosparam>
      <rosparam param="points_seen_topics"> ["points_seen_left", "points_seen_right"] </rosparam>

      <param name="use_masks"         type="bool" value="true"/>

      <remap from="~camera_left" to="/$(arg uav_name)/uvdar_bluefox_left/image_raw"/>
      <remap from="~camera_right" to="/$(arg uav_name)/uvdar_bluefox_right/image_raw"/>
      <remap from="~points_seen_left" to="/$(arg uav_name)/uvdar/points_seen_left"/>
      <remap from="~points_seen_right" to="/$(arg uav_name)/uvdar/points_seen_right"/>
      <remap from="~points_seen_left/sun" to="/$(arg uav_name)/uvdar/points_seen_left/sun"/>
      <remap from="~points_seen_right/sun" to="/$(arg uav_name)/uvdar/points_seen_right/sun"/>

      <remap from="~odometry" to="/$(arg uav_name)/mrs_odometry/new_odom"/>
      <remap from="~imu" to="mavros/imu/data"/>
    </node>

    <node name="blink_processor" pkg="nodelet" type="nodelet" args="$(arg nodelet) uvdar/UVDARBlinkProcessor $(arg nodelet_manager)" output="screen" respawn="false"> 
    <!-- <node name="blink_processor" pkg="nodelet" type="nodelet" args="$(arg nodelet) uvdar/UVDARBlinkProcessor $(arg nodelet_manager)" output="screen" respawn="false" launch-prefix="roslaunch_debug"> --> 
      <param name="uav_name" type = "string" value="$(arg uav_name)"/>
      <param name="debug"		type="bool" value="$(arg debug)"/> 
      <param name="visual_debug"		type="bool" value="$(arg visual_debug)"/> 
      <param name="gui"             type="bool" value="$(arg gui)"/>
      <param name="publish_visualization"             type="bool" value="true"/>
      <param name="use_camera_for_visualization"         type="bool" value="false"/>
      <param name="visualization_rate"         type="double" value="2.0"/>

      <rosparam param="frequencies" subst_value="true"> [$(arg frequency1), $(arg frequency2), $(arg frequency3), $(arg frequency4)] </rosparam>

      <param name="blink_process_rate"         type="int" value="$(arg blink_process_rate)"/>

      <param name="accumulator_length"         type="int" value="$(arg accumulator_length)"/>
      <param name="pitch_steps"         type="int" value="$(arg pitch_steps)"/>
      <param name="yaw_steps"         type="int" value="$(arg yaw_steps)"/>
      <param name="max_pixel_shift"         type="int" value="$(arg max_pixel_shift)"/>
      <param name="nullify_radius"         type="int" value="8"/>

      <rosparam param="camera_topics"> ["camera_left", "camera_right"] </rosparam>
      <rosparam param="points_seen_topics"> ["points_seen_left", "points_seen_right"] </rosparam>
      <rosparam param="blinkers_seen_topics"> ["blinkers_seen_left", "blinkers_seen_right"] </rosparam>
      <rosparam param="estimated_framerate_topics"> ["estimated_framerate_left", "estimated_framerate_right"] </rosparam>

      <remap from="~camera_left" to="/$(arg uav_name)/uvdar_bluefox_left/image_raw"/>
      <remap from="~camera_right" to="/$(arg uav_name)/uvdar_bluefox_right/image_raw"/>
      <remap from="~points_seen_left" to="/$(arg uav_name)/uvdar/points_seen_left"/>
      <remap from="~points_seen_right" to="/$(arg uav_name)/uvdar/points_seen_right"/>
      <remap from="~blinkers_seen_left" to="/$(arg uav_name)/uvdar/blinkers_seen_left"/>
      <remap from="~blinkers_seen_right" to="/$(arg uav_name)/uvdar/blinkers_seen_right"/>
      <remap from="~estimated_framerate_left" to="/$(arg uav_name)/uvdar/estimated_framerate_left"/>
      <remap from="~estimated_framerate_right" to="/$(arg uav_name)/uvdar/estimated_framerate_right"/>
      <remap from="~visualization" to="/$(arg uav_name)/uvdar/blink_visualization/image_raw"/>

    </node>
      
    <!-- <node name="uvdar_uav_pose_calculator_node" pkg="uvdar_core" type="uvdar_uav_pose_calculator_node" output="screen" launch-prefix="debug_roslaunch"> --> 
    <node name="uvdar_uav_pose_calculator_node" pkg="uvdar_core" type="uvdar_uav_pose_calculator_node" output="screen" > 
      <param name="uav_name" type = "string" value="$(arg uav_name)"/>
      <param name="debug"		type="bool" value="false"/> 
      <param name="gui"             type="bool" value="false"/>
      <param name="publish_visualization"             type="bool" value="true"/>
      <param name="frequencies_per_target" type="int" value="1"/>
      <rosparam param="frequencies" subst_value="true"> [$(arg frequency1), $(arg frequency2), $(arg frequency3), $(arg frequency4)] </rosparam>

      <param name="quadrotor" type="bool" value="true"/>
      <param name="beacon" type="bool" value="$(arg beacon)"/>
      <param name="arm_length" type="double" value="0.245"/>

      <rosparam param="blinkers_seen_topics"> ["blinkers_seen_left", "blinkers_seen_right"] </rosparam>
      <rosparam param="estimated_framerate_topics"> ["estimated_framerate_left", "estimated_framerate_right"] </rosparam>
      <rosparam param="camera_frames" subst_value="true"> ["$(arg uav_name)/uvcam_left", "$(arg uav_name)/uvcam_right"] </rosparam>

      <remap from="~blinkers_seen_left" to="/$(arg uav_name)/uvdar/blinkers_seen_left"/>
      <remap from="~blinkers_seen_right" to="/$(arg uav_name)/uvdar/blinkers_seen_right"/>
      <remap from="~estimated_framerate_left" to="/$(arg uav_name)/uvdar/estimated_framerate_left"/>
      <remap from="~estimated_framerate_right" to="/$(arg uav_name)/uvdar/estimated_framerate_right"/>

      <rosparam param="calib_files" subst_value="True"> ["default", "default"] </rosparam>

      <remap from="~measuredPoses1" to="/$(arg uav_name)/uvdar/measuredPosesL"/>
      <remap from="~measuredPoses2" to="/$(arg uav_name)/uvdar/measuredPosesR"/>
    </node>
  </group>


</launch>
